csharp/src/Ice/OutputStream-Class.cs

```diff
@@ -311,37 +354,139 @@ private void WriteInstance(AnyClass v)
+            {
+                if (compactId is int compactIdValue)
+                {
+                    typeIdKind = EncodingDefinitions.TypeIdKind.CompactId11;
+                    WriteSize(compactIdValue);
+                }
+                else
+                {
+                    int index = RegisterTypeId(typeId);
+                    if (index < 0)
+                    {
+                        typeIdKind = EncodingDefinitions.TypeIdKind.String;
+                        WriteString(typeId);
+                    }
+                    else
+                    {
+                        typeIdKind = EncodingDefinitions.TypeIdKind.Index;
+                        WriteSize(index);
+                    }
+                }
             }
+            else
+            {
+                Debug.Assert(compactId == null);
+                // With the 1.1 encoding, we always write a string and don't set a type ID kind in SliceFlags.
+                WriteString(typeId);
+            }
+
+            _current.SliceFlags |= (EncodingDefinitions.SliceFlags)typeIdKind;
         }
 
-        private sealed class InstanceData
+        /// <summary>Writes the type ID or type ID sequence immediately after the slice flags byte of the first slice,
+        /// and updates the slice flags byte as needed. Applies formal type optimization (class only), if possible.
+        /// </summary>
+        /// <param name="allTypeIds">The type IDs of all slices of this class or exception instance.</param>
+        /// <param name="errorMessage">The exception's error message. Provided only for exceptions.</param>
+        private void WriteTypeId20(string[] allTypeIds, string? errorMessage)
         {
-            internal readonly InstanceType InstanceType;
+            Debug.Assert(_current.InstanceType != InstanceType.None);
+
+            EncodingDefinitions.TypeIdKind typeIdKind = EncodingDefinitions.TypeIdKind.None;
 
+            if (_current.InstanceType == InstanceType.Class)
+            {
+                string typeId = allTypeIds[0];
+                if (typeId != _current.FormalTypeId20)
+                {
+                    int index = RegisterTypeId(typeId);
+                    if (index < 0)
+                    {
+                        if (_format == FormatType.Sliced)
+                        {
+                            typeIdKind = EncodingDefinitions.TypeIdKind.Sequence20;
+                            WriteSequence(allTypeIds, IceWriterFromString);
+                        }
+                        else
+                        {
+                            typeIdKind = EncodingDefinitions.TypeIdKind.String;
+                            WriteString(typeId);
+                        }
+                    }
+                    else
+                    {
+                        typeIdKind = EncodingDefinitions.TypeIdKind.Index;
+                        WriteSize(index);
```

Comment:
maybe call directly `WriteVarULong` as this method is only for 2.0 encoding?

Reply:
Same as for ReadSize above.